### PR TITLE
Add Job Board to PyCon Backend

### PR DESCRIPTION
**Pull request by patrick91**
This PR adds a Job Board, since we need as we sell it to our sponsors. I've decided to add it our main backend instead of creating a new service to keep it simple, we can make a new service in future 😊

This is based on our previous job board and how the blog posts are structured, there's some code duplication which we should remove (maybe in a future PR).

This PR only adds the backend for the job board, see tests to see how it is implemented.

As for blog posts, we should do pagination, but we can do that in future :)

### CODE REVIEW

File: backend/api/job_board/schema.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                slug=listing.slug,
+                description=listing.description,
+                company=listing.company,
+                company_logo=(
+                    info.context.request.build_absolute_uri(listing.company_logo.url)
+                    if listing.company_logo
+                    else None
+                ),
+                apply_url=listing.apply_url,
+            )
```
Comment (patrick91) [inline backend/api/job_board/schema.py:27]: do we have a new pattern for these kind of things? are we fully migrating to DDD?

Comment (marcoacierno) [reply backend/api/job_board/schema.py:26]: This is what I did in another PR: https://github.com/pythonitalia/pycon/pull/2571/files#diff-d65bfc90e5a4a3392ce87e3c0eab864a0b05b1cd7ec8a860a9803a975beba1dfR65

File: backend/job_board/models.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        filters = Q()
+
+        for lang, __ in settings.LANGUAGES:
+            filters |= Q(**{f"slug__{lang}": term})
+
+        return self.get_queryset().filter(filters)
+
+
+class JobListing(TimeStampedModel):
```
Comment (marcoacierno) [inline backend/job_board/models.py:25]: should we add OrderedModel so we can order the job listings?

File: backend/api/tests/test_job_board.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        "description": str(listing.description),
+        "company": str(listing.company),
+        "companyLogo": get_image_url_from_request(request, listing.company_logo),
+        "applyUrl": str(listing.apply_url),
+    } == resp["data"]["jobListings"][0]
+
+
+@mark.django_db
+def test_query_single_job_listing(rf, graphql_client, job_listing_factory):
+    request = rf.get("/")
```
Comment (estyxx) [inline backend/api/tests/test_job_board.py:51]: `request` is not used in this test, can we remove it?

File: backend/api/tests/test_job_board.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            jobListing(slug: "slug") {
+                title(language: "it")
+                slug(language: "it")
+            }
+        } """
+    )
+
+    assert not resp.get("errors")
+    assert resp["data"]["jobListing"] == {
+        "title": "diventa una lumaca",
```
Comment (estyxx) [inline backend/api/tests/test_job_board.py:110]: 😂

File: backend/job_board/models.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    title = I18nCharField(_("title"), max_length=200)
+    slug = I18nCharField(_("slug"), max_length=200, blank=True)
+    company = models.CharField(_("company"), max_length=100)
+    company_logo = models.ImageField(
+        _("company logo"), null=True, blank=True, upload_to="job-listings"
+    )
+    description = I18nTextField(_("description"), blank=True)
+    apply_url = models.URLField(_("URL where you can apply"), blank=True)
+
+    objects = JobListingManager()
```
Comment (estyxx) [inline backend/job_board/models.py:35]: The job offers are not related to the conference at tall? We will just delete it every year I suppose then...

File: backend/job_board/admin.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+from django.contrib import admin
+
+from .models import JobListing
+
+
+@admin.register(JobListing)
+class PostAdmin(admin.ModelAdmin):
+    model = JobListing
+    list_display = ("title", "company")
```
Comment (estyxx) [inline backend/job_board/admin.py:9]: can we add some filters and searchers?

File: backend/job_board/models.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    class Meta:
+        ordering = ["created"]
+
+    def __str__(self):
+        return f"[{self.company}] - {self.title}"
+
+    def save(self, *args, **kwargs):
+        if not self.slug:
+            self.slug = copy(self.title)
+            self.slug.map(slugify)
```
Comment (estyxx) [inline backend/job_board/models.py:46]:
```suggestion
            self.slug = slugify(self.title)
```
Why not just use the `slugify` django's function?

File: backend/job_board/models.py (hunk 2 above)
Comment (patrick91) [reply backend/job_board/models.py:25]: I think they should be ordered by creation date, we could add some additional flags like "highlighted" to move them at the top if needed, but we don't need custom order for now :)

File: backend/job_board/models.py (hunk 5 above)
Comment (patrick91) [reply backend/job_board/models.py:35]: I've asked Matteo, even about things like expiration and so on :)

We don't need anything complex so for now a global listing page is fine, we can think about improvements in future :D

File: backend/job_board/admin.py (hunk 6 above)
Comment (patrick91) [reply backend/job_board/admin.py:9]: there's not going to be a lot of jobs :)

File: backend/job_board/models.py (hunk 7 above)
Comment (patrick91) [reply backend/job_board/models.py:46]: because slug is a dictionary :)

File: backend/job_board/models.py (hunk 2 above)
Comment (marcoacierno) [reply backend/job_board/models.py:25]: So we won't have cases where a sponsor who paid 5k wants their job to appear at the start and not below those who paid 500 euros or whatever?

Fine if we don't want to add it for now

File: backend/api/job_board/schema.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+from .types import JobListing as JobListingType
+
+
+@strawberry.type
+class JobBoardQuery:
+    @strawberry.field
+    def job_listings(self, info) -> List[JobListingType]:
+        return [
+            JobListingType(
```
Comment (marcoacierno) [inline backend/api/job_board/schema.py:14]: I think we can avoid duplicating this init logic (here and in job_listing detail) with the `from_django_model`: https://github.com/pythonitalia/pycon/pull/2571/files#diff-d65bfc90e5a4a3392ce87e3c0eab864a0b05b1cd7ec8a860a9803a975beba1dfR65

File: backend/job_board/models.py (hunk 7 above)
Comment (marcoacierno) [reply backend/job_board/models.py:46]: TIL LazyI18nString has `map` :D had no idea, was confused why this worked just to find out that we have a map function in the class

File: backend/api/job_board/types.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+@strawberry.type
+class JobListing:
+    id: strawberry.ID
+    title: str = strawberry.field(resolver=make_localized_resolver("title"))
+    slug: str = strawberry.field(resolver=make_localized_resolver("slug"))
+    description: str = strawberry.field(resolver=make_localized_resolver("description"))
+    company: str
+    company_logo: Optional[str]
+    apply_url: Optional[str]
```
Comment (marcoacierno) [inline backend/api/job_board/types.py:17]:
```suggestion
    apply_url: str
```

I don't think this should be null, if it isn't specified empty string

File: backend/api/job_board/types.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+    def __init__(
+        self,
+        id: str,
+        title: str,
+        slug: str,
+        description: str,
+        company: str,
+        company_logo: Optional[str],
+        apply_url: Optional[str],
```
Comment (marcoacierno) [inline backend/api/job_board/types.py:27]:
```suggestion
        apply_url: str,
```

**Review by marcoacierno — APPROVED** [review]
I left a comment about using `from_django_model` so to remove the code-duplication in the strawberry type and another about a type that doesn't seem right.

Overall it looks good :)

**Pull request merged**